src/Heatmap/HeatMapInterpolationPoints.swift

```diff
@@ -0,0 +1,340 @@
+* limitations under the License.
+*/
+
+import Foundation
+import GooglePlaces
+import GoogleMaps
+import GoogleMapsUtils
+
+/// This class will create artificial points in surrounding locations with appropriate intensities interpolated by neighboring intensity values.
+class HeatMapInterpolationPoints {
+
+    /// The input data set
+    private var data = [GMUWeightedLatLng]()
+
+    /// The heat map layer
+    private let heatMapLayer: GMUHeatmapTileLayer = GMUHeatmapTileLayer()
+
+    /// The heat map colors and gradient
+    private var gradientColors = [UIColor.green, UIColor.red]
+    private var gradientStartPoints = [0.005, 0.7] as [NSNumber]
+
+    /// The list of interpolated heat map points with weight
+    private var heatMapPoints = [GMUWeightedLatLng]()
+
+    /// Indicates the number of times k-means clustering should execute; will be set in the constructor to 25 by default
+    private var clusterIterations: Int!
+
+    /// Normalizing factors to convert from 2D to longitude and latitude
+    private let normalLat = 175.9783070993
+    private let normalLong = 180.0
+
+    /// The constructor to this class
+    ///
+    /// - Parameter givenClusterIterations: The number of iterations k-means clustering should go to.
+    init(givenClusterIterations: Int = 25) {
+        clusterIterations = givenClusterIterations
+    }
+
+    // MARK: Functions that parse given data needed to build an interpolated heat map from
+
+    /// Adds a list of GMUWeightedLatLng objects to the input data set
+    ///
+    /// - Parameter latlngs: The list of GMUWeightedLatLng objects to add.
+    public func addWeightedLatLngs(latlngs: [GMUWeightedLatLng]) {
+        data.append(contentsOf: latlngs)
+    }
+
+    /// Adds a single GMUWeightedLatLng object to the input data set
+    ///
+    /// - Parameter latlngs: The list of GMUWeightedLatLng objects to add.
+    public func addWeightedLatLng(latlng: GMUWeightedLatLng) {
+        data.append(latlng)
+    }
+
+    /// Removes all previously supplied GMUWeightedLatLng objects
+    public func removeAllData() {
+        data.removeAll()
+    }
+
+    // MARK: Functions that directly contribute to the creation of interpolated points
+
+    /// A helper function that calculates the straight-line distance between two coordinates
+    ///
+    /// - Parameters:
+    ///   - lat1: The latitude value of the first point.
+    ///   - long1: The longitude value of the second point.
+    ///   - lat2: The latitude value of the second point.
+    ///   - long2: The longitude value of the second point.
+    /// - Returns: A double value representing the distance between the given points.
+    private func distance(lat1: Double, long1: Double, lat2: Double, long2: Double) -> Double {
+
+        // The normalizing factor scales down the spherical distance to a value that is more
+        // consistent with the parameters of the formulas
+        let normalizingFactor = 111195.0837241998
```

Comment:
How was this determined?

Reply:
I found this source: https://en.wikipedia.org/wiki/Decimal_degrees. I wanted to double check to see if it translates correctly, so I manually calculated the lat/long distance and compared that to the distance returned by meters of the GMSGeometryDistance function. The source said around 111.32 kilometers per degree, but I found 111.95 kilometers to be slightly more accurate. I've made a note of this in the code.

Is this method of conversion ok?